### PR TITLE
CTO-67 - Implemented Stream, StreamLocation and StreamJunctionIdentifier DTOs

### PR DESCRIPTION
**Pull request by rma-bryson**

### CODE REVIEW

File: cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
+public final class Stream implements CwmsDTOBase {
+
+    private final Boolean startsDownstream;
+    private final String flowsIntoStream;
+    private final Double flowsIntoStation;
+    private final String flowsIntoBank;
+    private final String divertsFromStream;
+    private final Double divertsFromStation;
+    private final String divertsFromBank;
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java]: From a JSON perspective, it is hard to determine why some of these are String (ids) vs Double (flows). Recommending to rename for clarity. Also `divertsFromBank` can likely be a L/R (left/right) enum

File: cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        result = 31 * result + Objects.hashCode(getDivertsFromStation());
+        result = 31 * result + Objects.hashCode(getDivertsFromBank());
+        result = 31 * result + Objects.hashCode(getLength());
+        result = 31 * result + Objects.hashCode(getSlope());
+        result = 31 * result + Objects.hashCode(getComment());
+        result = 31 * result + Objects.hashCode(getLocationIdentifier());
+        return result;
+    }
+
+    public static class Builder {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java]: should also be final

File: cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java (hunk 1 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java]: This has been re-done to use a new stream junction identifier object

File: cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java (hunk 2 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/Stream.java]: made final

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    private StreamJunctionIdentifier(Builder builder) {
+        this.streamId = builder.streamId;
+        this.bank = builder.bank;
+        this.station = builder.station;
+    }
+
+    @Override
+    public void validate() throws FieldException {
+        if (this.streamId == null) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:61]: bank isn't required?

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Objects;
+
+@FormattableWith(contentType = Formats.JSON, formatter = JsonV1.class)
+@JsonDeserialize(builder = StreamJunctionIdentifier.Builder.class)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
+/**
+ * Reference object for a stream location containing the location identifier, bank, and station
+ */
+public final class StreamJunctionIdentifier implements CwmsDTOBase {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:47]: `station` isn't part of the uniqueness identifier, maybe just call it `StreamJunction`?

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    private final Double station;
+
+    private StreamJunctionIdentifier(Builder builder) {
+        this.streamId = builder.streamId;
+        this.bank = builder.bank;
+        this.station = builder.station;
+    }
+
+    @Override
+    public void validate() throws FieldException {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:60]: validate should call `streamId.validate()` after the null check

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        this.publishedStation = builder.publishedStation;
+        this.navigationStation = builder.navigationStation;
+        this.lowestMeasurableStage = builder.lowestMeasurableStage;
+        this.totalDrainageArea = builder.totalDrainageArea;
+        this.ungagedDrainageArea = builder.ungagedDrainageArea;
+    }
+
+    @Override
+    public void validate() throws FieldException {
+        if (this.streamLocationId == null) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java:66]: Should also call `streamLocationId.validate()`

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java (hunk 4 above)
Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:47]: or `StreamJunctionRef`? if that doesn't make sense then no need to rename

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java (hunk 3 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:61]: Doesn't look required in the db. Seems pretty useless without that information though, other than to say this stream intersects this other stream someplace on some side.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java (hunk 5 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:60]: added validate call

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java (hunk 6 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java:66]: added validate call

**Review by adamkorynta — APPROVED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+import cwms.cda.formatters.annotations.FormattableWith;
+import cwms.cda.formatters.json.JsonV1;
+
+import java.util.Objects;
+
+@FormattableWith(contentType = Formats.JSON, formatter = JsonV1.class)
+@JsonDeserialize(builder = Stream.Builder.class)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
+public final class Stream implements CwmsDTOBase {
```
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java:21]: Why is the physical location data omitted from these implementations?

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    public String getComment() {
+        return comment;
+    }
+
+    public LocationIdentifier getStreamId() {
+        return streamId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java:83]: I have questions on the addition of equals/hashcode in the DTOs in general.
1. They do not follow the business rules defined for this data's uniqueness.
2. They add more data than what is used for this data's uniqueness.
3. Is there a usage of hashcode/equals outside of unit tests?

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.Objects;
+
+@FormattableWith(contentType = Formats.JSON, formatter = JsonV1.class)
+@JsonDeserialize(builder = StreamJunctionIdentifier.Builder.class)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
+/**
+ * Reference object for a stream location containing the location identifier, bank, and station
+ */
+public final class StreamJunctionIdentifier implements CwmsDTOBase {
```
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:47]: This class is confusing, a junction is where two streams meet. This seems to be a stream node. Also, what is the purpose of this class?

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream-junction-identifier.json [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{
+  "stream-id": {
+    "office-id": "Office123",
+    "location-id": "Stream123"
```
Comment (rma-psmorris) [inline cwms-data-api/src/test/resources/cwms/cda/data/dto/stream-junction-identifier.json:4]: these examples are confusing as I think these should be stream locations and not stream identifiers

**Review by rma-psmorris — COMMENTED** [review]
I'd like to go over this code and JSON in a meeting with you.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java (hunk 7 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java:21]: A Location isn't returned from the db like in other cases such as Embankments. I think we want to use the minimum number of queries here, and let the client make the Location query if needed. Unless you think that a fully-qualified Location is always going to be needed here.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java (hunk 8 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java:83]: Removed these. This is addressed in this PR: https://github.com/USACE/cwms-data-api/pull/721

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java (hunk 9 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamJunctionIdentifier.java:47]: Renamed to StreamNode here: https://github.com/USACE/cwms-data-api/pull/721

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream-junction-identifier.json (hunk 10 above)
Comment (rma-bryson) [reply cwms-data-api/src/test/resources/cwms/cda/data/dto/stream-junction-identifier.json:4]: Looking at this again, this is the CwmsIdentifier that represents the stream that this stream-node is on.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java (hunk 7 above)
Comment (rma-psmorris) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java:21]: I think your reasoning is fine, end application can query the web service for the location physical data if it needs it

Comment (MikeNeilson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java:21]: Is it common that the application *doesn't need it* or that it already has it.

While minimizing database queries is good, so is minimizing CDA queries. For CDA we de-normalize as appropriate (which maybe not here, hence the follow up question.)

**Pull request merged**